FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,54 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					let localFeed = cache.localFeed
+					let timestamp = cache.timestamp
+					completion(.found(feed: localFeed, timestamp: timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try? ManagedCache.find(in: context).map(context.delete)
```

Comment:
If `ManagedCache.find` throws any errors, they should not be ignored with optional `try?`. Otherwise, we may leave the store in an invalid state with multiple caches, if the `delete` failed, but `insert` works.
```suggestion
				try ManagedCache.find(in: context).map(context.delete)
```